qiita_db/handlers/oauth2.py

```diff
@@ -0,0 +1,280 @@
+    ----------
+    client_id : str
+        The client ID to validate
+    client_secret : str, optional
+        The client secret code. If not given, will only validate if client_id
+        exists
+
+    Returns
+    -------
+    bool
+        Whether client_id exists and, if client_secret given, the secret
+        matches the id
+    """
+    with qdb.sql_connection.TRN:
+        sql = """SELECT EXISTS(
+                    SELECT *
+                    FROM qiita.oauth_identifiers
+                    WHERE client_id = %s{0})"""
+        sql_info = [client_id]
+        if client_secret is not None:
+            sql = sql.format("AND client_secret = %s")
+            sql_info.append(client_secret)
+        else:
+            sql = sql.format("AND client_secret IS NULL")
+        qdb.sql_connection.TRN.add(sql, sql_info)
+        return qdb.sql_connection.TRN.execute_fetchlast()
+
+
+def authenticate_oauth(f):
+    """Decorate methods with this to require valid Oauth2 Authorization header
+
+    If a valid header is given, teh handof is odne and the page is rendered. If
+    an invalid header is given, the json error message is automatically sent.
+    """
+    @functools.wraps(f)
+    def wrapper(self, *args, **kwargs):
+        def oauth_error(self, error_msg, error):
+            self.set_status(400)
+            self.write({'error': error,
+                        'error_description': error_msg})
+            self.finish()
+
+        header = self.request.headers.get('Authorization', None)
+        if header is None:
+            oauth_error(self, 'Oauth2 error: invalid access token',
+                        'invalid_request')
+            return
+        token_info = header.split()
+        if len(token_info) != 2 or token_info[0] != 'Bearer':
+            oauth_error(self, 'Oauth2 error: invalid access token',
+                        'invalid_grant')
+            return
+
+        token = token_info[1]
+        db_token = r_client.hgetall(token)
+        if not db_token:
+            # token has timed out or never existed
+            oauth_error(self, 'Oauth2 error: token has timed out',
+                        'invalid_grant')
+            return
+        # Check daily rate limit for key if password style key
+        user = db_token.get('user', None)
+        if user:
+            if not self.check_rate_limit(db_token['client_id'], user):
+                oauth_error(self, 'Oauth2 error: daily request limit reached',
+                            'invalid_grant')
+                return
+
+        return f(self, *args, **kwargs)
+    return wrapper
+
+
+class OauthBaseHandler(RequestHandler):
+    def write_error(self, status_code, **kwargs):
+        """Overriding the default write error in tornado RequestHandler"""
+        if status_code in {403, 404, 405}:
+            # We don't need to log these failues in the logging table
+            return
+        # log the error
+        from traceback import format_exception
```

Comment:
If this is copy-paste, shouldn't this point out that we are doing something wrong? i.e. we are duplicating code...